src/shogun/io/openml/OpenMLData.cpp

```diff
@@ -0,0 +1,271 @@
+	return_string = reader.get("data_features", "json", id);
+	document.Parse(return_string.c_str());
+	check_response(document, "data_features");
+	const Value& dataset_features = document["data_features"];
+	for (const auto& param : dataset_features["feature"].GetArray())
+	{
+		std::unordered_map<std::string, std::vector<std::string>> param_map;
+		for (const auto& param_descriptors : param.GetObject())
+		{
+			std::vector<std::string> second;
+			if (param_descriptors.value.IsArray())
+				for (const auto& v : param_descriptors.value.GetArray())
+					second.emplace_back(v.GetString());
+			else
+				second.emplace_back(param_descriptors.value.GetString());
+
+			param_map.emplace(param_descriptors.name.GetString(), second);
+		}
+		param_vector.push_back(param_map);
+	}
+
+	// qualities
+	std::vector<std::unordered_map<std::string, std::string>> qualities_vector;
+	return_string = reader.get("data_qualities", "json", id);
+	document.Parse(return_string.c_str());
+	check_response(document, "data_qualities");
+	const Value& data_qualities = document["data_qualities"];
+	for (const auto& param : data_qualities["quality"].GetArray())
+	{
+		std::unordered_map<std::string, std::string> param_map;
+		for (const auto& param_quality : param.GetObject())
+		{
+			if (param_quality.name.IsString() && param_quality.value.IsString())
+				param_map.emplace(
+				    param_quality.name.GetString(),
+				    param_quality.value.GetString());
+			else if (param_quality.name.IsString())
+				param_map.emplace(param_quality.name.GetString(), "");
+		}
+		qualities_vector.push_back(param_map);
+	}
+
+	auto result = std::make_shared<OpenMLData>(
+	    name, description, data_format, dataset_id, version, creator,
+	    contributor, collection_date, upload_date, language, licence, url,
+	    default_target_attribute, row_id_attribute, ignore_attribute,
+	    version_label, citation, tags, visibility, original_data_url, paper_url,
+	    update_comment, md5_checksum, param_vector, qualities_vector);
+	result->set_api_key(api_key);
+	return result;
+}
+
+std::shared_ptr<CFeatures> OpenMLData::get_features() noexcept
+{
+	if (!m_cached_features)
+		get_data();
+	return m_cached_features;
+}
+
+std::shared_ptr<CFeatures> OpenMLData::get_features(const std::string& label)
+{
+	if (!m_cached_features)
+		get_data();
+	auto find_label =
+	    std::find(m_feature_names.begin(), m_feature_names.end(), label);
+	if (find_label == m_feature_names.end())
+		SG_SERROR("Requested label \"%s\" not in the dataset!\n", label.c_str())
+	auto col_idx = std::distance(m_feature_names.begin(), find_label);
+	auto feat_type_copy = m_feature_types;
+	feat_type_copy.erase(feat_type_copy.begin() + col_idx);
+	for (const auto type : feat_type_copy)
+	{
+		if (type == ARFFDeserializer::Attribute::STRING)
+			SG_SERROR("Currently cannot process string features!\n")
+	}
+	//	auto result = std::make_shared<CDenseFeatures>();
+	std::shared_ptr<CDenseFeatures<float64_t>> result;
```

Comment:
there is again the question of the datatype inside the feautres.....

Reply:
Yea, I am thinking if it would be giving the user the option of how to add a preprocessing step for this?

Reply:
sounds good!